d2/src/main/java/com/linkedin/d2/xds/IPv6AwarePickFirstLoadBalancer.java

```diff
@@ -0,0 +1,162 @@
+package com.linkedin.d2.xds;
+
+import io.grpc.EquivalentAddressGroup;
+import io.grpc.LoadBalancer;
+import io.grpc.LoadBalancerProvider;
+import io.grpc.LoadBalancerRegistry;
+import io.grpc.NameResolver.ConfigOrError;
+import io.grpc.Status;
+import io.grpc.internal.PickFirstLoadBalancerProvider;
+import java.net.Inet6Address;
+import java.net.InetSocketAddress;
+import java.net.SocketAddress;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * This {@link LoadBalancer} is an extension of the "pick_first" strategy which shuffles the addresses in a way that
+ * remains aware of whether the client supports IPv6. Namely, {@link #acceptResolvedAddresses} will be invoked with a
+ * list of addresses whose order represents its preference for IPv4 vs IPv6. For example, the list may contain only
+ * IPv6 addresses, only IPv4 addresses, IPv6 addresses followed by IPv4 addresses, IPv4 addresses followed by IPv6
+ * addresses, and everything in between. This load balancer shuffles the given addresses in a way that respects the
+ * original interleaving of address types, such that the client's preference is respected, while still achieving the
+ * desired effect of random pick_first load balancing.
+ */
+public class IPv6AwarePickFirstLoadBalancer extends LoadBalancer
+{
+  static
+  {
+    LoadBalancerRegistry.getDefaultRegistry().register(new Provider());
+  }
+
+  public static final String POLICY_NAME = "ipv6_aware_random_pick_first";
```

Comment:
Nit: I think the policy names `PICK_FIRST` and `RANDOM_PICK_FIRST` we use today are all caps, do we want to match that here?

Reply:
Gotcha, this should be fine then